main/src/library/Chain.flix

```diff
@@ -0,0 +1,891 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// The Chain type.
+///
+/// A chain is a list represented as an unbalanced binary tree.
+/// It supports efficient append and "snoc" - appending elements at the tail
+/// of the list.
+///
+/// Note - the constructors `Empty`, `One` and `Chain` should not be used directly.
+///
+enum Chain[t] {
+    case Empty
+    case One(t)
+    case Chain(Chain[t], Chain[t])
+}
+
+instance ToString[Chain[a]] with ToString[a] {
+    pub def toString(xs: Chain[a]): String =
+        let fn = () -> {
+            let sb = StringBuilder.new();
+            StringBuilder.appendString!(sb, "Chain#{");
+            Chain.foldLeft((ac, x) -> {
+                if (ac < 1)
+                    StringBuilder.appendString!(sb, "${x}")
+                else
+                    StringBuilder.appendString!(sb, ", ${x}");
+                ac + 1}, 0, xs);
+            StringBuilder.appendString!(sb, "}");
+            StringBuilder.toString(sb)
+        };
+        fn() as & Pure
+}
+
+instance Eq[Chain[a]] with Eq[a] {
+    pub def eq(xs: Chain[a], ys: Chain[a]): Bool = Chain.equals(xs, ys)
+}
+
+instance SemiGroup[Chain[a]] {
+    pub def combine(xs: Chain[a], ys: Chain[a]): Chain[a] = Chain.append(xs, ys)
+}
+
+instance Monoid[Chain[a]] {
+    pub def empty(): Chain[a] = Chain.empty()
+    pub def combine(xs: Chain[a], ys: Chain[a]): Chain[a] = Chain.append(xs, ys)
+}
+
+instance Functor[Chain] {
+    pub def map(f: a -> b & ef, xs: Chain[a]): Chain[b] & ef = Chain.map(f, xs)
+}
+
+instance Applicative[Chain] {
+    pub def point(x: a) : Chain[a] = Chain.point(x)
+    pub def ap(f: Chain[a -> b & ef], x: Chain[a]) : Chain[b] & ef = Chain.ap(f, x)
+}
+
+instance Monad[Chain] {
+    pub def flatMap(f: a -> Chain[b] & ef, x : Chain[a]) : Chain[b] & ef = Chain.flatMap(f, x)
+}
+
+instance Foldable[Chain] {
+    pub def foldLeft(f: (b, a) -> b & ef, s: b, xs: Chain[a]): b & ef = Chain.foldLeft(f, s, xs)
+    pub def foldRight(f: (a, b) -> b & ef, s: b, xs: Chain[a]): b & ef = Chain.foldRight(f, s, xs)
+}
+
+
+
+
+namespace Chain {
+
+    ///
+    /// A datatype for pattern matching on a chain (traversing left-to-right).
+    ///
+    pub enum ViewLeft[t] {
+        case NoneLeft
+        case SomeLeft(t, Chain[t])
+    }
+
+    instance Eq[ViewLeft[a]] with Eq[a] {
+        pub def eq(xs: ViewLeft[a], ys: ViewLeft[a]): Bool = match (xs, ys) {
+            case (NoneLeft, NoneLeft)                   => true
+            case (SomeLeft(x1, xs1), SomeLeft(y1, ys1)) => x1 == y1 and xs1 == ys1
+            case (_, _)                                 => false
+        }
+    }
+
+    ///
+    /// A datatype for pattern matching on a chain (traversing right-to-left).
+    ///
+    pub enum ViewRight[t] {
+        case NoneRight
+        case SomeRight(Chain[t], t)
+    }
+
+    instance Eq[ViewRight[a]] with Eq[a] {
```

Comment:
It looks like these instances can be derived from the enum if you want:

```scala
pub enum ViewRight[t] with Eq {
  ...
```

Reply:
Changed